ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/MLSdkAsyncHttpResponseHandler.java

```diff
@@ -0,0 +1,221 @@
+import software.amazon.awssdk.http.SdkHttpFullResponse;
+import software.amazon.awssdk.http.SdkHttpResponse;
+import software.amazon.awssdk.http.async.SdkAsyncHttpResponseHandler;
+
+@Log4j2
+public class MLSdkAsyncHttpResponseHandler implements SdkAsyncHttpResponseHandler {
+    @Getter
+    private Integer statusCode;
+    @Getter
+    private final StringBuilder responseBody = new StringBuilder();
+
+    private WrappedCountDownLatch countDownLatch;
+
+    private ActionListener<List<ModelTensors>> actionListener;
+
+    private Map<String, String> parameters;
+
+    private Map<Integer, ModelTensors> tensorOutputs;
+
+    private Connector connector;
+
+    private ScriptService scriptService;
+
+    private final static Gson GSON = new GsonBuilder().disableHtmlEscaping().create();
+
+    public MLSdkAsyncHttpResponseHandler(
+        WrappedCountDownLatch countDownLatch,
+        ActionListener<List<ModelTensors>> actionListener,
+        Map<String, String> parameters,
+        Map<Integer, ModelTensors> tensorOutputs,
+        Connector connector,
+        ScriptService scriptService
+    ) {
+        this.countDownLatch = countDownLatch;
+        this.actionListener = actionListener;
+        this.parameters = parameters;
+        this.tensorOutputs = tensorOutputs;
+        this.connector = connector;
+        this.scriptService = scriptService;
+    }
+
+    @Override
+    public void onHeaders(SdkHttpResponse response) {
+        SdkHttpFullResponse sdkResponse = (SdkHttpFullResponse) response;
+        log.debug("received response headers: " + sdkResponse.headers());
+        this.statusCode = sdkResponse.statusCode();
+    }
+
+    @Override
+    public void onStream(Publisher<ByteBuffer> stream) {
+        stream.subscribe(new MLResponseSubscriber());
+    }
+
+    @Override
+    public void onError(Throwable error) {
+        log.error(error.getMessage(), error);
+        if (statusCode == null) {
+            actionListener
+                .onFailure(
+                    new OpenSearchStatusException(
+                        "Error on communication with remote model: " + error.getMessage(),
+                        RestStatus.INTERNAL_SERVER_ERROR
+                    )
+                );
+        } else {
+            actionListener
+                .onFailure(
+                    new OpenSearchStatusException(
+                        "Error on communication with remote model: " + error.getMessage(),
+                        RestStatus.fromCode(statusCode)
+                    )
+                );
+        }
+    }
+
+    private void processResponse(
+        Integer statusCode,
+        String body,
+        Map<String, String> parameters,
+        Map<Integer, ModelTensors> tensorOutputs
+    ) {
+        ModelTensors tensors;
+        if (Strings.isBlank(body)) {
+            log.error("Remote model response body is empty!");
+            tensors = processErrorResponse(statusCode, "null");
```

Comment:
"null" --> This will not make any sense to customers. What do you think about `Remote model response is empty!` as default message?

Reply:
Sounds good.